src/main/java/org/truffleruby/language/objects/classvariables/ResolveTargetModuleForClassVariablesNode.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.language.objects.classvariables;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.NodeChild;
+import com.oracle.truffle.api.dsl.Specialization;
+import org.truffleruby.core.module.RubyModule;
+import org.truffleruby.language.LexicalScope;
+import org.truffleruby.language.RubyContextSourceNode;
+import org.truffleruby.language.RubyGuards;
+import org.truffleruby.language.RubyNode;
+
+@NodeChild(value = "lexicalScopeNode", type = RubyNode.class)
+public abstract class ResolveTargetModuleForClassVariablesNode extends RubyContextSourceNode {
+
+    public static ResolveTargetModuleForClassVariablesNode create() {
+        return ResolveTargetModuleForClassVariablesNodeGen.create(null);
+    }
+
+    public abstract RubyModule execute(LexicalScope lexicalScope);
+
+    @Specialization(guards = "lexicalScope == cachedLexicalScope")
+    protected RubyModule cached(LexicalScope lexicalScope,
+            @Cached("lexicalScope") LexicalScope cachedLexicalScope,
+            @Cached("uncached(lexicalScope)") RubyModule cachedModule) {
+        return cachedModule;
+    }
+
+    @TruffleBoundary
```

Comment:
We don't really need a boundary here, that loop would PE fine (but unprofiled and with volatile reads).
OTOH, it's probably convenient to notice that case with `--engine.InstrumentBoundaries`, as cases where the LexicalScope is not constant at a given site should be rather rare.

Reply:
What I was going to do if we want to look at more indirect class variables is to have a node that handles `n` levels of indirection with an unrolled loop or something, so I wouldn't remove this boundary (which was also already there.)

Reply:
Yeah, we could inline cache how many levels up to look, indeed.
OK to keep the boundary.